### ISSUE TITLE
Performance: Xml parser allocates a String for every content character in the document.

### ISSUE THREAD

**Issue by corliss** [AngleSharp/AngleSharp#144]
From XmlDomBuilder.cs:

``` c#
                case XmlTokenType.Character:
                {
                    var tok = (XmlCharacterToken)token;
                    CurrentNode.AppendText(tok.Data.ToString());
                    break;
                }
```

This creates a 20-byte string for every character, so should be avoided, as this is the most significant code path in document parsing.

Incidentally, I recently came across this [Channel9 talk](http://channel9.msdn.com/Events/TechEd/NorthAmerica/2013/DEV-B333) about how the Roslyn team converted the C#/VB compilers to managed code. It has a lot of great information - including spotting allocations like the one in this issue. I recommend watching it to anyone interested in maximizing managed code performance.

Cheers,

**Comment by FlorianRappl** [comment]
This is due to the specification (please read it carefully). However, in the HTML tokenizer I made a slight perf. optimization, which combines many characters into one (HTML specification also demands spitting out a single token for each character). For XML, however, I did not do that (no focus on XML). In general it requires a lot more thought. It could / should be similar to the implementation in the HTML tokenizer, but since I can't spent time thinking about it, I won't optimize it.

Maybe somebody else wants to do it.
=> Marked as "up for grab".

Remark: The 20-bytes do not hurt (they are thrown away anyway). What hurts more is the GC pressure. I did not watch that specific talk, but I am sure they must have mentioned it. A string has a char[], hence ends up on the heap. If it would have been a struct there would be no problem (besides potentially requiring more bytes per parameter than a simple 8-byte (x64) reference).

**Comment by corliss** [comment]
Yes, the GC pressure is the main thing. The byte size and object count both matter - the byte count causes the pressure, the number of objects is the complexity of collection.

For users that are interested in parsing, not tokenization (there's currently no way to use the tokenizer since it's internal - I'm opening a separate request for that) this optimization is perfectly legal, correct? Therefore this optimization is quite critical - do you have benchmarks for XML parsing vs other parsers? I suspect that for large files, the slowdown will be significant just because of this one line.

[BTW I really recommend this talk - it's one that equals several years of experience. I've already watched it  a couple of times.]

**Comment by FlorianRappl** [comment]
Yes, but byte count doesn't matter that much in my experience (unless you break the LOH barrier). The allocation is really expensive. Not only because it needs to be tracked by the GC, but also context switching and synchronization barriers hurt.

The temporary string allocation you are after here could be solved temporarily by providing another method in `Node` (such as `AppendCharacter`), but even though it would alleviate the described situation it would not cure it. The cure is as described in the tokenizer. Then character tokens contain a string already and the `AppendCharacter` method would be obsolete again. That is why I see this as the root cause.

Unfortunately I have no timings on the XML parser, but I guess that what you say is perfectly right - large files (high character to tag ratio) should run slow(er) than the competition.

One remark (just because it came to my mind): A while ago (maybe a year) I tried around with token reusage (especially re-using the character tokens, but also tag tokens etc.). However, the result wasn't significantly better (it was even running slower for small files as far as I can remember). In the end the code readability was significantly worse and there were edge cases and unanswered questions. Hence I rejected it.

**Comment by corliss** [comment]
What would the problem be with having character tokens that have a character rather than a string? The token itself can be reused, hopefully.

**Comment by FlorianRappl** [comment]
You still need to create a token / character. Since the case of just a single character (e.g. between tag tokens) appearing is much less likely than multiple characters, optimizing for the latter makes sense.

**Comment by corliss** [comment]
I'm a bit confused by your answer.
For normal xml like `<tag>abcdef</tag>` currently the reason you stated that `char.ToString()` is used is that the character token takes a string, because tokens _need_ to be returned one char at a time, and that tokens have a string field, is this correct? If so I'm asking why not have tokens have a character field instead.

**Comment by FlorianRappl** [comment]
The property `Data` is exposing the `_data` character field. Maybe we are not talking about the same thing. The token _already_ has a character field.

What I was saying is that this particular character field should be a string. That string then needs to be used (if it is always filled with just a single character we are back at the original problem), i.e. we need to scan all characters and not just the next one.

**Comment by corliss** [comment]
OK, that makes sense. Agreed.

**Comment by FlorianRappl** [comment]
Alright I will take care of this. Will be shipped with AngleSharp v0.9.2.

**Comment by corliss** [comment]
Great. I will try out these fixes as soon as they are merged into master. (I'm not using nuget.)
